src/renderer/mesh_storage.zig

```diff
@@ -967,7 +971,153 @@ pub fn updateLightMap(map: *LightMap.LightMapFragment) void {
 
 // MARK: Block breaking animation
 
-pub fn addBreakingAnimation(pos: Vec3i, breakingProgress: f32) void {
+const BlockDamage = struct {
+	updateQueue: main.utils.ConcurrentQueue(Update),
+	damage: std.AutoHashMapUnmanaged(Vec3i, f32),
+	mutex: std.Thread.Mutex,
+	secondsSinceLastUpdate: f64,
+
+	const updateIntervalSeconds: f64 = 1.0;
+
+	const Update = union(enum) {
+		add: struct {pos: Vec3i, progress: f32},
+		remove: struct {pos: Vec3i},
+	};
+
+	fn init() BlockDamage {
+		return .{
+			.updateQueue = .init(main.globalAllocator, 16),
+			.damage = .{},
+			.mutex = .{},
+			.secondsSinceLastUpdate = 0,
+		};
+	}
+	fn deinit(self: *BlockDamage) void {
+		self.mutex.lock();
+		defer self.mutex.unlock();
+
+		self.updateQueue.deinit();
+		self.damage.deinit(main.globalAllocator.allocator);
+	}
+	fn clear(self: *BlockDamage) void {
+		main.utils.assertLocked(&self.mutex);
+
+		var iterator = self.damage.iterator();
+		while(iterator.next()) |entry| {
+			const pos = entry.key_ptr.*;
+			self.removeAnimation(pos);
+		}
+		self.damage.clearRetainingCapacity();
+	}
+	pub fn set(self: *BlockDamage, pos: Vec3i, remainingHealth: f32) void {
```

Comment:
Instead of set/get, shouldn't we use add/subtract to avoid having to do multiple hashmap lookups?